ldms/src/core/ldms_xprt.c

```diff
@@ -3764,20 +3764,27 @@ int ldms_xprt_sockaddr(ldms_t _x, struct sockaddr *local_sa,
 {
 	struct ldms_xprt *x = _x;
 	zap_err_t zerr;
- again:
-	if (x->sa_len) {
-		/* use cached addresses */
-		*sa_len = x->sa_len;
-		memcpy(local_sa, &x->local_sa, x->sa_len);
-		memcpy(remote_sa, &x->remote_sa, x->sa_len);
-		return 0;
-	}
-	zerr = zap_get_name(x->zap_ep, &x->local_sa, &x->remote_sa, &x->sa_len);
-	if (zerr) {
-		x->sa_len = 0;
-		return -1;
+	socklen_t __sa_len = *sa_len;
+
+	if (*sa_len == 0)
+		return EINVAL;
+
+	if (!x->sa_len) {
+		zerr = zap_get_name(x->zap_ep, &x->local_sa,
+					&x->remote_sa, &__sa_len);
+		if (zerr)
+			return zap_zerr2errno(zerr);
+		x->sa_len = __sa_len;
 	}
-	goto again;
+
+	/*
+	 * Truncate the address if the given buffers are smaller
```

Comment:
This is an EINVAL. If this is the case, then the address written is garbage, truncating it avoids the buffer overflow, but the address written is invalid.

Reply:
@tom95858 I'm aware that truncated address is garbage. I'm mimic what `getsockname()` does. `getsockname()` doesn't return an error if the input buffer is smaller than the actual address size. I'll return EINVAL instead then.